File: tools/sinf_finish_case/hasher.py
from pathlib import Path
import config
import markers
import hashlib
from portable import put_portable
from logger import get_logger
from tqdm import tqdm
import multiprocessing


def sha512(path):
    path = Path(path)
    hash_sha512 = hashlib.sha512()
    with path.open("rb") as f:
        for chunk in iter(lambda: f.read(4096), b""):
            hash_sha512.update(chunk)
    return hash_sha512.hexdigest()


def calc_hash(path):
    path = Path(path)
    try:
        text = sha512(path)
        return path, text, None
    except (FileNotFoundError, OSError) as e:
        return path, "", e


class Hasher:

    def __init__(self, folder):
        self.f = None
        self.n_files = 0
        self.root = Path(folder)
        self.hash_file = self.root / "hash.txt"
        self.pbar = None
        self.logger = get_logger(self.root / "hash.log")
        self.n_workers = 4


    def check_file(self, path: Path):
        if path.name == ".sinf_mark.json":
            return False
        return True
       

    @staticmethod
    def __is_hash_partial(path: Path):
        markers_ = markers.get_markers_folder(path)
        for marker in markers_:
            if marker and marker['type'] == "hash_partial":
                return marker['subtype']

    def hash_normal_folder(self, folder: Path):
        for entry in folder.iterdir():
            if entry.is_dir():
                for item in self.hash_folder(entry):
                    yield item
            else:
                if self.check_file(entry):
                    yield entry
      

    def hash_folder_iped_images(self, folder: Path):
        for item in folder.iterdir():
            if item.is_file() and item.suffix in [".log", ".txt"]:
                if self.check_file(item):
                    yield item
            elif item.is_dir():
                for item in self.hash_folder(item):
                    yield item
    
      
    def hash_folder_iped_results(self, folder: Path):
        put_portable(folder)
        item = folder / "FileList.csv"
        if not item.exists():
            item = item / "Lista de Arquivos.csv"
        if item.exists():
            if self.check_file(item):
                yield item
           
        

    def hash_folder(self, folder: Path):
        type_ = self.__is_hash_partial(folder)
        if type_ == "iped_results":
            for item in self.hash_folder_iped_results(folder):
                yield item
        elif type_ == "iped_images":
            for item in self.hash_folder_iped_images(folder):
                yield item
        else:
            for item in self.hash_normal_folder(folder):
                yield item


    def count_files(self):
        print("Contando arquivos")
        self.n_files = 0
        for item in self.hash_folder(self.root):
            self.n_files += 1
        print(f"{self.n_files} arquivos encontrados")



    def run(self):
        self.count_files()
        self.logger.info("Inciando")
        try:
            self.f = self.hash_file.open("w", encoding="utf-8")
            self.pbar = tqdm(total=self.n_files)
            pool = multiprocessing.Pool(processes=self.n_workers)
            for path, hash_, err in pool.imap(calc_hash, self.hash_folder(self.root)):
                if err:
                    self.logger.error(str(err))
                else:
                    self.f.write(f"{hash}     {path.relative_to(self.root)}\n")
                self.pbar.update(1)
            self.pbar.close()
        finally:
            self.f.close()
        res = sha512(self.hash_file)
        print("Hash do hash")
        print(res)
        self.logger.info("Finalizado")
        self.logger.info(f"Hash do hash: {res}")
        


if __name__ == "__main__":
    hasher = Hasher('.')
    hasher.run()

File: tools/sinf_finish_case/go/hasher.go
package main

import (
	"crypto/sha512"
	"encoding/hex"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"path/filepath"
	"sync"

	"github.com/cheggaaa/pb/v3"
	"github.com/google/logger"
)

//Hasher df
type Hasher struct {
	// f         *os.File
	onlyCount bool
	nFiles    int64
	root      string
	hashFile  string
	// pBar      *pb.ProgressBar
	nWorkers int
	jobs     chan string
	results  chan Result
	wg       sync.WaitGroup
}

//Result is
type Result struct {
	Path     string
	HashText string
	Error    error
}

func calculateSha512(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()

	h := sha512.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}

	hasher := sha512.New()
	s, err := ioutil.ReadFile(path)
	hasher.Write(s)
	if err != nil {
		return "", err
	}
	return hex.EncodeToString(hasher.Sum(nil)), nil
}

func (hasher *Hasher) calcHashFile(path string) {
	if filepath.Base(path) == ".sinf_mark.json" {
		return
	}
	if hasher.onlyCount {
		hasher.nFiles++
		return
	}
	hasher.jobs <- path

}

func (hasher *Hasher) hashNormalFolder(path string) {
	items, _ := ioutil.ReadDir(path)
	for _, item := range items {
		if item.IsDir() {
			hasher.hashFolder(filepath.Join(path, item.Name()))
		} else {

			hasher.calcHashFile(filepath.Join(path, item.Name()))
		}
	}
}

func (hasher *Hasher) hashFolderIpedImages(path string) {
	items, _ := ioutil.ReadDir(path)
	for _, item := range items {
		if item.IsDir() {
			hasher.hashFolder(filepath.Join(path, item.Name()))
		} else {
			switch filepath.Ext(item.Name()) {
			case ".log", ".txt":
				hasher.calcHashFile(filepath.Join(path, item.Name()))
			}
		}
	}
}

func (hasher *Hasher) hashFolderIpedResults(path string) {
	item := filepath.Join(path, "FileList.csv")
	if !fileExists(item) {
		item = filepath.Join(path, "Lista de Arquivos.csv")
	}
	if fileExists(item) {
		hasher.calcHashFile(item)
	}
	putPortable(path)

}

func (hasher *Hasher) hashFolder(path string) {
	ok, marker := isHashPartial(path)
	if ok {
		switch marker.Subtype {
		case "iped_results":
			hasher.hashFolderIpedResults(path)
		case "iped_images":
			hasher.hashFolderIpedImages(path)
		}
	} else {
		hasher.hashNormalFolder(path)
	}
}

func (hasher *Hasher) countFiles() {
	fmt.Println("Contando arquivos. Aguarde, isso pode demorar.")
	hasher.onlyCount = true
	hasher.hashFolder(hasher.root)
	hasher.onlyCount = false
	fmt.Printf("%d arquivos encontrados\n", hasher.nFiles)
}

func (hasher *Hasher) workerHash() {
	defer func() {
		hasher.results <- Result{HashText: "finish"}
	}()

	for path := range hasher.jobs {
		// if path == "finish" {
		// 	return
		// }
		result := Result{}
		text, err := calculateSha512(path)
		if err != nil {
			result.Path, _ = filepath.Rel(hasher.root, path)
			result.HashText = ""
			result.Error = err
		} else {
			relPath, _ := filepath.Rel(hasher.root, path)
			result.Path = relPath
			result.HashText = text
		}
		hasher.results <- result
	}
	fmt.Println("Finalizou worker")
}

func (hasher *Hasher) run() {
	// var wg sync.WaitGroup
	hasher.jobs = make(chan string)
	hasher.results = make(chan Result)
	hasher.countFiles()
	logger.Info("Iniciando.")
	f, err := os.OpenFile(hasher.hashFile, os.O_CREATE|os.O_WRONLY, 0660)
	if err != nil {
		logger.Fatalf("Failed to open hash file: %v", err)
	}
	defer f.Close()
	pBar := pb.Full.Start64(hasher.nFiles)

	// hasher.wg.Add(hasher.nWorkers)
	for i := 0; i < hasher.nWorkers; i++ {
		go hasher.workerHash()
	}

	go func() {
		hasher.hashFolder(hasher.root)
		close(hasher.jobs)
	}()

	nRunning := hasher.nWorkers
	for result := range hasher.results {
		if result.Error != nil {
			logger.Errorf("Não foi possível calcular hash do arquivo \"%s\"", result.Path)
		} else if result.HashText == "finish" {
			nRunning--
			if nRunning == 0 {
				break
			}
			continue
		} else {
			fmt.Fprintf(f, "%s	%s\n", result.HashText, result.Path)
		}
		pBar.Increment()
	}

	// hasher.wg.Wait()
	pBar.Finish()
	res, err := calculateSha512(hasher.hashFile)
	if err != nil {
		logger.Fatalf("Failed to open hash file: %v", err)
	}
	fmt.Printf("Hash do hash: \n%s", res)
	logger.Infof("Finalizado")
	logger.Infof("Hash do hash: %s", res)
}
